src/merkle_tree.rs

```diff
@@ -184,24 +203,40 @@ impl MerkleTree {
 }
 
 /// Verifies a given merkle proof is valid.
-pub fn verify(
-    element: String,
+pub fn verify_tree_element<T>(tree: &MerkleTree, element: T, proof: Vec<String>) -> bool
+where
+    T: Hash,
+{
+    let mut hasher = DefaultHasher::new();
+    element.hash(&mut hasher);
+    let hashed_element = hasher.finish().to_string();
+    let root_hash = tree.get_root_hash();
+    let elem_index = tree.leaf_index_of(&hashed_element);
+
+    if let Some(index) = elem_index {
+        verify(hashed_element, index, proof, root_hash)
+    } else {
+        false
+    }
+}
+
+fn verify(
+    mut hashed_element: String,
```

Comment:
we could make these parameters non mutable to make this a pure function

Reply:
Does something like this seem like a good idea?

```
fn verify(
    hashed_element: &String,
    elem_index: usize,
    proof: Vec<String>,
    root_hash: String,
) -> bool {
    // create local mutable copies
    let mut hashed_element = hashed_element.clone();
    let mut elem_index = elem_index;
    //...
}
```

Or maybe taking ownership of `hashed_element` and calling the function with element.clone()

Reply:
Seems good to me

Reply:
done